src/container/oid_iv.c

```diff
@@ -85,15 +86,21 @@ oid_iv_ent_refresh(struct ds_iv_entry *iv_entry, struct ds_iv_key *key,
 
 	D_ASSERT(priv);
 	num_oids = priv->num_oids;
-	D_DEBUG(DB_MD, "%u: ON REFRESH %zu\n", dss_self_rank(), num_oids);
+	D_DEBUG(DB_MD, "%u: ON REFRESH: num_oids = %zu, REF_RC = " DF_RC "\n", dss_self_rank(),
+		num_oids, DP_RC(ref_rc));
 	D_ASSERT(num_oids != 0);
 
 	entry = iv_entry->iv_value.sg_iovs[0].iov_buf;
 	D_ASSERT(entry != NULL);
 
 	/** if iv op failed, just release the entry lock acquired in update */
-	if (ref_rc != 0)
+	if (ref_rc != 0) {
+		/** if retrying request, we do not increase oids */
+		entry->current_req_retry = true;
```

Comment:
1) Yes i considered that and it does not matter actually. if an op is not retryable, the error is just going to return to the client and the request discarded anyway.
2) I already modified the debug message on line 89 that include ref_rc. that is how i discovered the issue.

for your suggestion, I think it makes it more complicated to expand oid_iv_range IMO: a) it's something that is used outside of this IV (in contianer_target) and we are adding IV tree specific items to it. b) it's used for allocation but also tracking local OIDs available, so overloading it with those things IMO is not appropriate.

Reply:
"if an op is not retryable, the error is just going to return to the client and the request discarded anyway."
I am not sure what the "request discarded" mean, for the same pool and container, the "current_req_retry" flag will be left in the persistent IV entry and affect following oid_iv_reserve() operation. It is not fatal but not clean.

another point is for example if the ref_rc is returned by root node, then the current_req_retry will be left on each node through the path from root to the request node (may with middle node) when the refresh() callback called. Then the retried reserve request, from request node to root node, each node will reserve same number of OIDs, that is different with the original intention that parent node reserve more so next request need not go to root again. Right? Again it is not fatal but looks not clean to me.

Reply:
> another point is for example if the ref_rc is returned by root node, then the current_req_retry will be left on each node through the path from root to the request node (may with middle node) when the refresh() callback called.
OK good point. I can change the check to set it only on retryable RPC which should resolve that concern.

> By using similar method as my last comment, the retired request will keep the same behavior that parent node reserve more OIDs same as before. Or I may misunderstood sth?
I think your suggestion really overloads the range data-structure. for each hop you need to update the req_rank also. to me it sounds more complicated than just a flag for retry operations.

Reply:
thanks for updating the RC check. There are still a some differences of retry case with _iv_op(), i.e. a few detailed case still will left the current_req_retry flag for non-retriable case.
I think it is better to avoid do that kind of check if possible.

"for each hop you need to update the req_rank also"
only need to set the req_rank in oid_iv_reserve() before call ds_iv_update(). With that info on each node's oid_iv_ent_update() it have a chance to check if self_rank == req_rank in that case it get the original req_num_oids rather than already increased "oids->num_oids" to avoid overflow when retry.
That can remove the current_req_retry flag and keep the same behavior of reserving more OIDs at parent node same as non-retried case.
Please consider.
I can +1 first as this PR is an improvement anyway.